### ISSUE TITLE
Fetch message with a given Message Sequence Number

### ISSUE THREAD

**Issue by sphh** [ikvk/imap_tools#249]
I track messages with their Message Sequence Numbers (not UIDs), because the IMAP server delivers them when in IDLE mode.

I know how to fetch a message, if I know its UID. But how do I fetch the message, if I only have its Message Sequence Number?

Thanks.

**Comment by ikvk** [comment]
You may see it at old versions, less 1.1.0

**Comment by ikvk** [comment]
See mailbox.py at 1.0.0 

def _fetch_by_one
def _fetch_in_bulk
def fetch

Now numbers() are not using in fetch()

**Comment by sphh** [comment]
I see. Thanks.

Can this feature re-added again? Either by a dedicated `mailbox` method or with a switch in these three methods (`uid_or_number='uid'`?)

**Comment by ikvk** [comment]
Why are you need it? What is your purpose?

**Comment by sphh** [comment]
I watch a folder with `IDLE`. `IDLE` responses have a message sequence number (and not an UID) after the `*`. To fetch that message, I would either need the UID (which is not sent by the `IDLE` command) or I need a `mailbox` method fetching the message with that message sequence number.

Is there another way to fetch the message given by the `IDLE` response?

**Comment by ikvk** [comment]
May you show me ar RFC that IDLE responce contains numbers? 
https://datatracker.ietf.org/doc/html/rfc2177

Here is IDLE examples
https://github.com/ikvk/imap_tools/blob/master/examples/idle.py

**Comment by sphh** [comment]
> May you show me at RFC that IDLE responce contains numbers and not uids? https://datatracker.ietf.org/doc/html/rfc2177

That RFC says:
> […] the server is now free to send untagged EXISTS, EXPUNGE, and other messages at any time

Looking up these messages in [RFC2060, 7.4 Message Status](https://datatracker.ietf.org/doc/html/rfc2060#section-7.4) reveals, that
> Immediately following the "*" token is a number that represents a message sequence number.

This is a _message sequence number_ and not the message's UID. Is that what you are looking for in the RFC?

> Here is IDLE examples https://github.com/ikvk/imap_tools/blob/master/examples/idle.py

I can always ask for everything from the server, but to reduce traffic, sometimes (that is, when receiving a `FETCH` from the server), it is sufficient to get the message mentioned with the message sequence number.

**Comment by ikvk** [comment]
Got it.
I think the effort is not worth it - Many code for very rare usage. People get confused even with the current API.
Why just not use AND(seen=False) query on IDLE response?

**Comment by ikvk** [comment]
> … yesterday I switched to one of your ‘competitor’ because of this.

You say as if I forbid you to write custom methods for BaseMailBox :D 
It would be better to suggest MR. 

I will investigate this feature more close. I hope I'll find the time.

**Comment by ikvk** [comment]
> convert message sequence numbers to message uids

 - good idea for api

**Comment by ikvk** [comment]
Check it:
https://github.com/ikvk/imap_tools/releases/tag/v1.8.0

Thanks. 
And say hi to Menno.

**Comment by sphh** [comment]
Wow, that was quick. I am impressed! Maybe I have to switch back this package :thinking:

Thanks @ikvk and Menno!

**Comment by ikvk** [comment]
Show me your magic package

**Comment by ikvk** [comment]
I would appreciate for a practical example of using numbers_to_uids.

**Comment by sphh** [comment]
Nothing to present yet. But I can outline the idea: All biff programs I have seen so far display the unread messages. None of them can distinguish between the unread messages and the recently arrived. For example, the biff program I have in mind shows a window with the list of new messages. The window is dismissed with a click on it. The next time a new message arrives, this window should only show this new message – and not all (other) unread messages. And if a message is marked as unread in the mail reader, this message should also not be shown as new message.

And since this is a long-running process, it should use `IDLE` instead of polling. I could keep track of the message sequence numbers, but unfortunately they seem to change, hence I have to use the message UIDs.

The responses in `IDLE` inform my program of changes in the message status. But only with the message sequence number and not the UID. That's why I need to convert those message sequence numbers to UIDs.

The alternative is to keep new, unseen and all UIDs in three separate lists. I then have to fetch the UIDS for the unseen and for all messages for every single change and update those lists. Using the status changes as described above would probably require only two lists (new and unseen) and would definitely remove a lot of requests to the server.

**Comment by ikvk** [comment]
Thanks.
Concurent lib is imapclient?

**Comment by sphh** [comment]
Yes. :innocent: When I noticed, that `imap_tools` could not get hold of the UIDs I was looking for an alternative. `imapclient` cannot do it either but its API turned out to be more accessible for me (which is a personal thing). But maybe I have to move back, now that `imap_tools` grew this feature? :thinking:

**Comment by ikvk** [comment]
to be honest, I don't understand why imapclient is needed, it doesn't make the developer's life easier.
